consul/consul.go

```diff
@@ -137,15 +134,15 @@ func (c *Consul) Register(service *registry.Service) {
 //
 func (c *Consul) Deregister() error {
 	for s, b := range serviceCache {
-		if !b.isRegistered {
+		if c.CacheIsValid(s) {
+			c.CacheProcessDeregister(s)
+		} else {
```

Comment:
I'm missing something then. How is it getting deregistered the first time? The first time through `CacheIsValid` is going to return true since the validityCounter isn't incremented until `c.CacheProcessDeregister` is run.